### ISSUE TITLE
Vocabulary overflow Issue with [PAD] for SFT

### ISSUE THREAD

**Issue by MagicDevilZhang** [OpenLLMAI/OpenLLaMA2#87]
When expanding the tokenizer's vocabulary with `[PAD]` during SFT (https://github.com/OpenLLMAI/OpenLLaMA2/blob/main/examples/utils.py#L22), the tokenizer's vocabulary size becomes 32001 (e.g., `print(size(tokenzier.vocab))`). 

![image](https://github.com/OpenLLMAI/OpenLLaMA2/assets/43229014/e3164b39-db1e-47ea-8817-f11d1043dcc1)

However, the model's embedding can only handle up to 32000 entries.

![image](https://github.com/OpenLLMAI/OpenLLaMA2/assets/43229014/34d79ee7-89a5-452f-bcb4-0256db69b7c7)

This issue arises apparent when the startup parameter `args.micro_train_batch_size=2` is utilized. 

The default `args.micro_train_batch_size=1`, the dataloader does not invoke `dataset.collect_fn`, which consequently bypasses the padding process (and the potential overflow of the [PAD] token).

I would appreciate attention to this matter as it affects cases where larger batch sizes are used during SFT.

**Comment by MagicDevilZhang** [comment]
```
deepspeed ./examples/train_sft.py \
    --max_len 2048 \
    --dataset 'Open-Orca/OpenOrca' \
    --dataset_probs '1.0' \
    --train_batch_size 128 \
    --micro_train_batch_size 2 \
    --max_samples 100 \
    --pretrain "meta-llama/Llama-2-7b-hf" \
    --save_path "./ckpt/7b_llama" \
    --zero_stage 2 \
    --max_epochs 1 \
    --bf16 \
    --learning_rate 5e-6 \
    --gradient_checkpointing
```

**Comment by MagicDevilZhang** [comment]
Please refer as : https://github.com/huggingface/transformers/issues/22312

**Comment by hijkzzz** [comment]
Thanks for your findings, I will try to fix the problem.

**Comment by hijkzzz** [comment]
Fixed
